### ISSUE TITLE
Architecture and SIMD agnostic

### ISSUE THREAD

**Issue by michel2323** [exanauts/ExaPF.jl#7]
Right now, `kernel.jl` implements a CPU and GPU abstraction. It takes care of either generating CUDA kernels or just a regular function to run on the CPU. There are also other details that make target = "cpu"/"cuda" work. Macros are widely used and the way it is done now allows only for one instantiation of the either CUDA or CPU code.

This should be revisited with another more flexible extraction. Also, what is decided here has an effect of how a framework for nonlinear equations or optimization would like.

**Comment by frapac** [comment]
I investigate more this issue. I think at some point we should be able to find a better abstraction model in `ExaPF.jl`. Thinking more about it, I am wondering if we could add a dependency to [KernelAbstractions.jl](https://github.com/JuliaGPU/KernelAbstractions.jl). The idea would be to define the code we are deporting to GPU in a `@kernel` macro, and then dispatch to CPU/GPU with a dedicated statement, e.g.
```julia
 if isa(a, Array)
     kernel! = powerflow_kernel!(CPU(), 4)
 else
     kernel! = powerflow_kernel!(CUDADevice(), 256)
 end
 kernel!(pf_inputs, ndrange=a_size_for_pf) 
```
(this snippet is adapted from [this example](https://juliagpu.gitlab.io/KernelAbstractions.jl/examples/matmul/)).

However, I am wondering whether `KernelAbstractions` is mature enough. Maybe we could use the JuliaCon as an opportunity to discuss with the developers?

**Comment by frapac** [comment]
Solved by #15
